rs/replicated_state/src/canister_state/queues.rs

```diff
@@ -344,6 +343,261 @@ impl From<RequestOrResponse> for CanisterInput {
+    fn queue_front_not_stale<T>(
+        &self,
+        queue: &CanisterQueue<T>,
+        canister_id: &CanisterId,
+    ) -> Result<(), String>
+    where
+        MessageStoreImpl: MessageStore<T>,
+    {
+        if let Some(reference) = queue.peek() {
+            if self.is_stale(reference) {
+                return Err(format!(
+                    "Stale reference at the front of {:?} queue to/from {}",
+                    reference.context(),
+                    canister_id
+                ));
+            }
+        }
+
+        Ok(())
+    }
+}
+
+/// Defines context-specific (inbound / outbound) message store operations
+/// (lookup, removal, staleness check) for `MessageStoreImpl`.
+trait MessageStore<T> {
+    /// The type returned by `get()`: `&T` if the implementation actually holds
+    /// items of type `T`; or the type `T` if it has to be built on demand.
+    type TRef<'a>
+    where
+        Self: 'a;
+
+    /// Looks up the referenced item. Panics if the reference is stale.
+    fn get(&self, reference: message_pool::Reference<T>) -> Self::TRef<'_>;
+
+    /// Removes the referenced item. Panics if the reference is stale.
+    fn take(&mut self, reference: message_pool::Reference<T>) -> T;
+
+    /// Checks whether the given reference is stale (i.e. neither in the pool, nor
+    /// in one of the compact response maps iff inbound).
+    fn is_stale(&self, reference: message_pool::Reference<T>) -> bool;
+}
+
+impl MessageStore<CanisterInput> for MessageStoreImpl {
+    type TRef<'a> = CanisterInput;
+
+    fn get(&self, reference: InboundReference) -> CanisterInput {
+        assert_eq!(Context::Inbound, reference.context());
+
+        if let Some(msg) = self.pool.get(reference) {
+            debug_assert!(!self.shed_responses.contains_key(&reference));
+            return msg.clone().into();
+        } else if reference.kind() == Kind::Response && reference.class() == Class::BestEffort {
+            if let Some(callback_id) = self.shed_responses.get(&reference) {
+                return CanisterInput::ResponseDropped(*callback_id);
+            }
+        }
+
+        panic!("stale reference at the front of input queue");
+    }
+
+    fn take(&mut self, reference: InboundReference) -> CanisterInput {
+        assert_eq!(Context::Inbound, reference.context());
+
+        if let Some(msg) = self.pool.take(reference) {
+            return msg.into();
+        } else if reference.kind() == Kind::Response && reference.class() == Class::BestEffort {
+            if let Some(callback_id) = self.shed_responses.remove(&reference) {
+                return CanisterInput::ResponseDropped(callback_id);
+            }
+        }
+
+        panic!("stale reference at the front of input queue");
+    }
+
+    fn is_stale(&self, reference: InboundReference) -> bool {
+        assert_eq!(Context::Inbound, reference.context());
+
+        self.pool.get(reference).is_none()
+            && (reference.context() != Context::Inbound
```

Comment:
I guess this is a leftover from before we had typed queues (the condition was `!Inbound || !Response || !shed", so the former was merely redundant).

But I'll just pretend I meant to check that the reference has `Class::BestEffort` (because `shed_responses` only ever holds inbound best-effort responses).